core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -341,9 +341,11 @@ class FileMessageSet private[kafka](@volatile var file: File,
     if(targetSize > originalSize || targetSize < 0)
       throw new KafkaException("Attempt to truncate log segment to " + targetSize + " bytes failed, " +
                                " size of this log segment is " + originalSize + " bytes.")
-    channel.truncate(targetSize)
-    channel.position(targetSize)
-    _size.set(targetSize)
+    if (targetSize != channel.size().toInt) {
```

Comment:
If we compare to originalSize then the testPreallocateClearShutdown will fail on the assert that the position did change on reopening the FileMessageSet and that the file is cleanly truncated. So in this case the size of the channel is much bigger (because it's preallocated) then the originalSize and targetSize. The javadoc of the tests says: "Test the new FileMessageSet with pre allocate as true and file has been clearly shut down, the file will be truncate to end of valid data."

More Details for failing Test:
FileMessageSetTest.scala:242 which is  assertEquals(oldposition, position)
Expected :114
Actual: 536870912